parlai/core/torch_agent.py

```diff
@@ -420,88 +420,119 @@ def add_cmdline_args(cls, argparser):
             default='sgd',
             choices=cls.optim_opts(),
             help='Choose between pytorch optimizers. Any member of torch.optim'
             ' should be valid.',
         )
-        agent.add_argument(
+        optim_group.add_argument(
             '-lr', '--learningrate', type=float, default=1, help='learning rate'
         )
-        agent.add_argument(
+        optim_group.add_argument(
             '-clip',
             '--gradient-clip',
             type=float,
             default=0.1,
             help='gradient clipping using l2 norm',
         )
-        agent.add_argument(
+        optim_group.add_argument(
             '-mom',
             '--momentum',
             default=0,
             type=float,
             help='if applicable, momentum value for optimizer.',
         )
-        agent.add_argument(
+        optim_group.add_argument(
             '--nesterov',
             default=True,
             type='bool',
             help='if applicable, whether to use nesterov momentum.',
         )
-        agent.add_argument(
+        optim_group.add_argument(
             '-nu',
             '--nus',
             default='0.7',
             type='floats',
             help='if applicable, nu value(s) for optimizer. can use a single '
             'value like 0.7 or a comma-separated tuple like 0.7,1.0',
         )
-        agent.add_argument(
+        optim_group.add_argument(
             '-beta',
             '--betas',
             default='0.9,0.999',
             type='floats',
             help='if applicable, beta value(s) for optimizer. can use a single '
             'value like 0.9 or a comma-separated tuple like 0.9,0.999',
         )
+        optim_group.add_argument(
+            '-wd',
+            '--weight-decay',
+            type=float,
+            default=None,
+            help='Weight decay on the weights.',
+        )
+
         # lr scheduler
-        agent.add_argument(
+        lr_group = agent.add_argument_group('Learning Rate Scheduler')
+        lr_group.add_argument(
             '--lr-scheduler',
             type=str,
             default='reduceonplateau',
-            choices=['reduceonplateau', 'none', 'fixed', 'invsqrt'],
+            choices=['reduceonplateau', 'none', 'fixed', 'invsqrt', 'cosine'],
             help='Learning rate scheduler.',
         )
-        agent.add_argument(
+        lr_group.add_argument(
             '--lr-scheduler-patience',
             type=int,
             default=3,
             help='LR scheduler patience. In number of validation runs. If using '
             'fixed scheduler, LR is decayed every <patience> validations.',
         )
-        agent.add_argument(
+        lr_group.add_argument(
             '--lr-scheduler-decay',
             type=float,
             default=0.5,
-            help='Decay factor for LR scheduler, or how much LR is multiplied by '
-            'when it is lowered.',
+            help=(
```

Comment:
do you need these parentheses

Reply:
No, my opinion was that it was just more readable this way. You prefer the other?

Reply:
doesn't really matter, other way looks a bit more consistent

Reply:
Sorry if stupid/intentional - is there a reason we don't incorporate their LR min value from the paper? (Seems like 0 is always used in the arg to the torch CosineAnnealingLR object). Too much trouble to add another config arg?

Reply:
Laziness tbh. I actually added it, it made a bug, and chose to delete the relevant code rather than figure out my bug. It's real unclear to me what's a right value for eta_min.

@glample you use 0? even though you said you only do a single period...

Reply:
Well, not a huge deal either way, but 0 seems a bit pointless no? The last few updates (esp the last one) before each restart are a waste in that case

Reply:
Only one is a waste, the rest are just REALLY tiny ;)